src/machi_cr_client.erl

```diff
@@ -548,12 +548,119 @@ do_read_chunk2(File, Offset, Size, Depth, STime, TO,
             read_repair(ConsistencyMode, read, File, Offset, Size, Depth, STime, S);
             %% {reply, {error, not_written}, S};
         {error, written} ->
-            exit({todo_should_never_happen,?MODULE,?LINE,File,Offset,Size})
+            exit({todo_should_never_happen,?MODULE,?LINE,File,Offset,Size});
+        {error, trimmed}=Err ->
+            {reply, Err, S}
+    end.
+
+do_trim_chunk(File, Offset, Size, 0=Depth, STime, TO, S) ->
+    do_trim_chunk(File, Offset, Size, Depth+1, STime, TO, S);
+    
+do_trim_chunk(File, Offset, Size, Depth, STime, TO, #state{proj=P}=S) ->
+    sleep_a_while(Depth),
+    DiffMs = timer:now_diff(os:timestamp(), STime) div 1000,
+    if DiffMs > TO ->
+            {reply, {error, partition}, S};
+       true ->
+            %% This is suboptimal for performance: there are some paths
+            %% through this point where our current projection is good
+            %% enough.  But we're going to try to keep the code as simple
+            %% as we can for now.
+            S2 = update_proj(S#state{proj=undefined, bad_proj=P}),
+            case S2#state.proj of
+                P2 when P2 == undefined orelse
+                        P2#projection_v1.upi == [] ->
+                    do_trim_chunk(File, Offset, Size, Depth + 1,
+                                  STime, TO, S2);
+                _ ->
+                    do_trim_chunk2(File, Offset, Size, Depth + 1,
+                                   STime, TO, S2)
+            end
+    end.
+
+do_trim_chunk2(File, Offset, Size, Depth, STime, TO,
+               #state{epoch_id=EpochID, proj=P, proxies_dict=PD}=S) ->
+    [HeadFLU|RestFLUs] = mutation_flus(P),
+    Proxy = orddict:fetch(HeadFLU, PD),
+    case ?FLU_PC:trim_chunk(Proxy, EpochID, File, Offset, Size, ?TIMEOUT) of
+        ok ->
+            %% From this point onward, we use the same code & logic path as
+            %% append does.
+            do_trim_midtail(RestFLUs, undefined, File, Offset, Size,
+                            [HeadFLU], 0, STime, TO, S);
+        {error, bad_checksum}=BadCS ->
+            {reply, BadCS, S};
+        {error, Retry}
+          when Retry == partition; Retry == bad_epoch; Retry == wedged ->
+            do_trim_chunk(File, Offset, Size, Depth, STime, TO, S);
+        {error, trimmed} ->
+            {reply, ok, S}
```

Comment:
Kota, please double-check, but I believe that we want to continue trimming the remaining middle & tail FLUs.  For example, if the chain is [A,B,C]:

* Client 1 attempts to trim a chunk at offset X.  The trim at A succeeds; no operation has been sent yet to B.
* Client 1 crashes.
* Client 2 wishes to trim the same chunk at offset X.
* Client 2 attempts to trim A and gets the `{error, trimmed}` response at thls line.
* Client 2 really ought to try to trim the chunk on servers B & C.